server/ante/bypass_ibc_fee_decorator.go

```diff
@@ -16,38 +18,60 @@ type anteHandler interface {
 
 type BypassIBCFeeDecorator struct {
 	nextAnte anteHandler
-	dk       distrkeeper.Keeper
-	sk       seqkeeper.Keeper
+	dk       distrKeeper
+	sk       sequencerKeeper
 }
 
-func NewBypassIBCFeeDecorator(nextAnte anteHandler, dk distrkeeper.Keeper, sk seqkeeper.Keeper) BypassIBCFeeDecorator {
+type distrKeeper interface {
+	GetPreviousProposerConsAddr(ctx sdk.Context) sdk.ConsAddress
+}
+
+type sequencerKeeper interface {
+	GetSequencerByConsAddr(ctx sdk.Context, consAddr sdk.ConsAddress) (stakingtypes.Validator, bool)
+	GetWhitelistedRelayers(ctx sdk.Context, operatorAddr sdk.ValAddress) (types.WhitelistedRelayers, error)
+}
+
+func NewBypassIBCFeeDecorator(nextAnte anteHandler, dk distrKeeper, sk sequencerKeeper) BypassIBCFeeDecorator {
 	return BypassIBCFeeDecorator{
 		nextAnte: nextAnte,
 		dk:       dk,
 		sk:       sk,
 	}
 }
 
-// AnteHandle skips fee deduct and min gas price ante handlers for whitelisted relayer messages
+// AnteHandle allows IBC relayer messages and skips fee deduct and min gas price ante handlers for whitelisted relayer messages
 func (n BypassIBCFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
-	// Check if the tx is from whitelisted relayer
-	whitelisted, err := n.isWhitelistedRelayer(ctx, tx.GetMsgs())
+	msgs := tx.GetMsgs()
+	var err error
+	msgs, err = n.getAllFinalMsgs(ctx, msgs, 0)
 	if err != nil {
-		// This error is not critical; just log and fall into the default fee handling
-		ctx.Logger().With("module", "BypassIBCFeeDecorator", "err", err).
-			Error("Failed to check if the tx is from the whitelisted relayer")
+		return ctx, err
 	}
-	if whitelisted {
-		// The tx is from the whitelisted relayer, so it's eligible for the fee exemption
-		return next(ctx, tx, simulate)
+
+	totalMsgs := len(msgs)
+	ibcCount := countIBCMsgs(msgs)
+
+	if ibcCount == totalMsgs {
+		// all are IBC messages
+		ibcWhitelisted, err := n.isIBCWhitelistedRelayer(ctx, msgs)
+		if err != nil {
+			return ctx, err
+		}
+		if ibcWhitelisted {
+			return next(ctx, tx, simulate)
+		}
+		return ctx, fmt.Errorf("not all signers whitelisted")
+	} else if ibcCount > 0 {
+		// mixed: some IBC and some non-IBC
+		return ctx, fmt.Errorf("mixed IBC and non-IBC messages in the same transaction not allowed")
 	}
 
 	// The tx if not from the whitelisted relayer; proceed with the default fee handling
 	return n.nextAnte.AnteHandle(ctx, tx, simulate, next)
 }
 
-// isWhitelistedRelayer checks if all the messages in the transaction are from whitelisted IBC relayer
-func (n BypassIBCFeeDecorator) isWhitelistedRelayer(ctx sdk.Context, msgs []sdk.Msg) (bool, error) {
+// isIBCWhitelistedRelayer checks if all the messages in the transaction are from whitelisted IBC relayer
+func (n BypassIBCFeeDecorator) isIBCWhitelistedRelayer(ctx sdk.Context, msgs []sdk.Msg) (bool, error) {
 	// Check if the tx is from IBC relayer
 	if !IsIBCRelayerMsg(msgs) {
```

Comment:
seems like we only enter this function if this condition is correct i.e `if ibcCount == totalMsgs` so not sure why should we check it twice.